virtual_rainforest/abiotic/radiation.py

```diff
@@ -0,0 +1,179 @@
+        of year if this is NOT provided as an input to the virtual rainforest.
+
+        Args:
+            n: day of the year
+            y: year
+            lat: latitude
+
+        Ref:
+            Davis et al. (2017): Simple process-led algorithms for simulating habitats
+            (SPLASH v.1.0): robust indices of radiation, evapotranspiration and plant-
+            available moisture, Geosci. Model Dev., 10, 689-708
+        """
+        raise NotImplementedError("Implementation of this feature is missing")
+
+    def calc_topofcanopy_radiation(
+        self,
+        sw_in: float,
+        n: int,
+        y: int,
+        lat: float,  # this is optional for the case of sw_in == FALSE
+        elev: float,
+        sf: float,
+        topo_factor: float,
+        sw_albedo: float,
+        vis_albedo: float,
+    ) -> None:
+        """Calculate top of canopy shortwave radiation.
+
+        If sw_in (extra-terrestrial) radiation is NOT provided as an input, it is
+        calculated as a function of latitude and date based on the SPLASH model
+        (Davis et al., 2017).
+
+        Args:
+            sw_in: shortwave downward radiation [W m-2]
+            n: day of year
+            y: year
+            lat: latitude
+            elev: elevation (m)
+            sf: fraction of sunshine hours
+            topo_factor: topographic adjustment factor for radiation (slope and aspect)
+            sw_albedo: shortwave albedo (function of vegetation and bare soil fraction)
+            vis_albedo: albedo visible light (for PPFD calculation)
+
+        Returns:
+            topofcanopy_radiation
+            ppfd
+        """
+        # check if shortwave down is provided, else calculate
+        if sw_in:
+            sw_down = sw_in
+        else:
+            sw_down = self.calc_sw_down(n, y, lat)
+
+        # atmospheric filtering (clouds, aerosols)
+        tau_o = kc + kd * sf
+        tau = tau_o * (1.0 + (2.67e-5) * elev)
+        sw_down_f = sw_down * tau
+
+        # topography and albedo effects on shortwave radiation
+        sw_net_toc = (1 - sw_albedo) * sw_down_f * topo_factor
+        self.sw_net_toc = sw_net_toc
+
+        # PPFD
+        # TODO integrate over duration of the day?
+        ppfd = (1.0e-6) * kfFEC * (1.0 - vis_albedo) * sw_down_f
+        self.ppfd = ppfd
+
+    def calc_canopy_radiation(
+        self, canopy_abs: float, tc: float, ts: float, emis: list
+    ) -> None:
+        """Calculates daily net radiation profile within the canopy.
+
+        The output is a vector of net radiation, the first entry is the top
+        of the canopy, the last entry is the net radiation that reaches the surface.
+
+        Args:
+            n_level: number of canopy layers
+            canopy_abs: absorption by canopy (from plant module) [W m-2]
+            emis: vector with canopy and soil emmissivity
+            tc: vector of canopy temperatures
```

Comment:
`n_level` is not in the function signature. Should this handle multiple cells or not (or not _now_). We should be using `numpy` types (or possibly `xarray` types) for vectors

Reply:
Agree. If the inputs are numpy arrays (or xarray.DataArrays) the type annotations should reflect that. The same in `radiation_balance` below. So, something like:

```python
from numpy.typing import NDArray
...
    def calc_canopy_radiation(
         self, canopy_abs: float, tc: NDArray[float], ts: float, emis: NDArray[float]
     ) -> None:
```

Also, probably you can get `n_level` (should it be `n_layers`?) from the length of `tc`, right?

Reply:
There are multiple dimensions, one is the cell_id and one is the vertical structure in each cell (n_layers) which can be derived from the length of tc, very good point, as it will change with the vegetation dynamics.
So far I have only thought about the types of the 'smallest unit' without taking space or time into account, I guess most inputs will be arrays, I will change that in the next iteration.

Reply:
I have a problem with NDArray[float] , it says: 'Could not specialize type "NDArray[ScalarType@NDArray]"
  Type "float" cannot be assigned to type "generic" "float" is incompatible with "generic"'. Suggestions?

Reply:
'emis' contains only two constant values, I added it to the list of constants at the top

Reply:
Or `NDArray[floating]`?

Reply:
Perfect, that seems to work. That's also what I concluded from my internet search but wasn't sure how that was compatible with what we have set up.

Reply:
The parallel issue I've run into is when a value could be a scalar or an array: `Union[float, np.ndarray]` . That causes `mypy` irritation down the line because you constantly have to trap where you can use e.g. `var[idx]` . I think another problem here is that 0-dimension np.arrays then get treated as scalars: `x = np.array(6.0)` ?

Is the best way to handle this to aggressively cast scalar types to one-dimensional np.arrays?

Reply:
Actually - maybe I'm looking for `npt.ArrayLike`, which should handle this?

Reply:
> The parallel issue I've run into is when a value could be a scalar or an array: `Union[float, np.ndarray]` . That causes `mypy` irritation down the line because you constantly have to trap where you can use e.g. `var[idx]` . I think another problem here is that 0-dimension np.arrays then get treated as scalars: `x = np.array(6.0)` ?
> 
> Is the best way to handle this to aggressively cast scalar types to one-dimensional np.arrays?

I think so, yes. Either that or just don't handle the scalar case and make the user convert it themselves. In general I imagine you won't be passing around scalar values a whole lot anyway, so I think that would be fine. Don't make your life harder than it needs to be.

I think `ArrayLike` is for other types that are like numpy arrays in certain respects, mainly in that you can convert them to a numpy array, so you could pass a tuple or a list into your function instead. I don't know though.

Reply:
Is this a n issue if I want to run a function for all cell_ids but not necessarily in parallel?

Reply:
@vgro Do you mean you have a function that can be applied using a vectorised calculation across all `cell_id`s? That isn't technically 'in parallel' as it is a single calculation but just applied to array inputs. Or that you don't want to do that and need to carry out calculations on a cell-by-cell basis?

Reply:
Could you rephrase your questions please, I'm not sure if I understand the terminology here. I want to do a single calculation but applied to array inputs. What type of processes would you run in parallel?

Reply:
By parallel, I meant code where you run each cell in a different process (like if you run a bunch of jobs on different cores on the HPC). If you were looping over cells, you could either do each in series using a `for` loop or farm out different cells to different processors and speed it up. As @alexdewar noted elsewhere here, that is something to think about later.

I've used 'vectorised' to mean applying the same calculation across all of the elements within arrays of inputs, which is something `numpy` is brilliant at, and is what you're describing here.

Reply:
That's useful to know, @dalonsoa . In my case, I was trying to support situations where someone might be working with a single site as opposed to calculating the same things across lots of locations. It would be better there to stipulate that users need to provide single sites using np.arrays, even if 0-dimensional.

Reply:
@vgro A good example here would be calculating pressure from elevation (assuming no _weather_ :smile:). If you had a `numpy` array of cell elevations from the data store, you could just do something like:

```python
data['air_pressure'] = 101325 *  (1 - 0.0000225577 * data['elevation']) ** 5.25588    
```

And that is all the cells done in one step